api/bin/chainflip-ingress-egress-tracker/src/main.rs

```diff
@@ -1,17 +1,145 @@
+			cf_chains::assets::eth::Asset::Usdc =>
+				Self { chain: ForeignChain::Ethereum, asset: asset.into() },
+		}
+	}
+}
+
+impl From<cf_chains::assets::dot::Asset> for WitnessAsset {
+	fn from(asset: cf_chains::assets::dot::Asset) -> Self {
+		match asset {
+			cf_chains::assets::dot::Asset::Dot =>
+				Self { chain: ForeignChain::Polkadot, asset: asset.into() },
+		}
+	}
+}
+
+impl From<cf_chains::assets::btc::Asset> for WitnessAsset {
+	fn from(asset: cf_chains::assets::btc::Asset) -> Self {
+		match asset {
+			cf_chains::assets::btc::Asset::Btc =>
+				Self { chain: ForeignChain::Bitcoin, asset: asset.into() },
+		}
+	}
+}
+
+#[derive(Serialize)]
+#[serde(tag = "type", rename_all = "snake_case")]
+enum TransactionId {
+	Bitcoin { hash: String },
+	Ethereum { signature: SchnorrVerificationComponents },
+	Polkadot { signature: String },
+}
+
+#[derive(Serialize)]
+#[serde(tag = "type", rename_all = "snake_case")]
+enum WitnessInformation {
+	Deposit {
+		src_chain_block_height: <AnyChain as Chain>::ChainBlockNumber,
+		deposit_address: String,
+		amount: NumberOrHex,
+		asset: WitnessAsset,
+	},
+	Broadcast {
+		broadcast_id: BroadcastId,
+		tx_out_id: TransactionId,
+	},
+}
+
+type EthereumDepositInfo = (DepositWitness<Ethereum>, <Ethereum as Chain>::ChainBlockNumber);
+
+impl From<EthereumDepositInfo> for WitnessInformation {
+	fn from((value, height): EthereumDepositInfo) -> Self {
+		Self::Deposit {
+			src_chain_block_height: height,
+			deposit_address: value.deposit_address.to_string(),
+			amount: value.amount.into(),
+			asset: value.asset.into(),
+		}
+	}
+}
+
+type BitcoinDepositInfo =
+	(DepositWitness<Bitcoin>, <Bitcoin as Chain>::ChainBlockNumber, BitcoinNetwork);
+
+impl From<BitcoinDepositInfo> for WitnessInformation {
+	fn from((value, height, network): BitcoinDepositInfo) -> Self {
+		Self::Deposit {
+			src_chain_block_height: height,
+			deposit_address: value.deposit_address.to_address(&network),
+			amount: value.amount.into(),
+			asset: value.asset.into(),
+		}
+	}
+}
+
+type PolkadotDepositInfo = (DepositWitness<Polkadot>, <Polkadot as Chain>::ChainBlockNumber);
+
+impl From<PolkadotDepositInfo> for WitnessInformation {
+	fn from((value, height): PolkadotDepositInfo) -> Self {
+		Self::Deposit {
+			src_chain_block_height: height as u64,
+			deposit_address: format!("0x{}", hex::encode(value.deposit_address.aliased_ref())),
+			amount: value.amount.into(),
+			asset: value.asset.into(),
+		}
+	}
+}
+
+async fn get_broadcast_id<I>(
+	state_chain_client: &StateChainClient<()>,
+	tx_out_id: &TransactionOutIdFor<state_chain_runtime::Runtime, I::Instance>,
+) -> Option<BroadcastId>
+where
+	state_chain_runtime::Runtime: pallet_cf_broadcast::Config<I::Instance>,
+	I: PalletInstanceAlias + 'static,
+{
+	state_chain_client
+		.storage_map_entry::<pallet_cf_broadcast::TransactionOutIdToBroadcastId<
+			state_chain_runtime::Runtime,
+			I::Instance,
+		>>(state_chain_client.latest_finalized_block().hash, tx_out_id)
```

Comment:
Have you tried this out? Think this is a bug. The ingress-egress tracker is using an unfinalised stream, so querying the finalised head might result in a "does not exist" type error, and return None here. I think best to use `latest_unfinalized_block_hash`

Reply:
Yeah we ran it and everything worked. I assume the finalized head will be behind the unfinalized head so it should always be able to find it?

Reply:
Changed it in two places

Reply:
Given the unfinalised is driving it, unfinalised can say "query this at block 20" but the finalised is at block 18, and so when querying at block 18, it might not be there

Reply:
Yeah i see what you mean. here it shouldn't be an issue because i don't think any broadcast happens within two blocks but fixed it nonetheless